src/ports/web/api/events/export.rs

```diff
@@ -0,0 +1,176 @@
+
+    let records: Vec<_> = events.map(adapters::csv::EventRecord::from).collect();
+
+    let buff: Vec<u8> = vec![];
+    let mut wtr = csv::Writer::from_writer(buff);
+
+    for r in records {
+        wtr.serialize(r)?;
+    }
+    wtr.flush()?;
+    let data = String::from_utf8(wtr.into_inner()?)?;
+
+    Ok(Content(ContentType::CSV, data))
+}
+
+#[get("/export/events.ical?<query..>")]
+pub fn ical_export(
+    connections: sqlite::Connections,
+    search_engine: tantivy::SearchEngine,
+    auth: Auth,
+    query: usecases::EventQuery,
+) -> result::Result<Content<String>, AppError> {
+    let events = query_events(connections, search_engine, auth, query)?;
+
+    use icalendar::{Calendar, Class, Component, Event, Property};
+
+    let mut calendar = events
+        .map(|ev| {
+            let mut e = Event::new()
+                .uid(ev.id.as_str())
+                .summary(&ev.title)
+                .starts(ev.start)
+                .class(Class::Public)
+                // https://icalendar.org/iCalendar-RFC-5545/3-8-1-2-categories.html
+                .append_property(Property::new("CATEGORIES", &ev.tags.join(",")).done())
+                .done();
+
+            if let Some(l) = &ev.location {
+                // https://icalendar.org/iCalendar-RFC-5545/3-8-1-6-geographic-position.html
+                e = e
+                    .append_property(
+                        Property::new(
+                            "GEO",
+                            &format!("{};{}", l.pos.lat().to_deg(), l.pos.lng().to_deg()),
+                        )
+                        .done(),
+                    )
+                    .done();
+
+                if let Some(addr) = address_from_location(l) {
+                    // https://icalendar.org/iCalendar-RFC-5545/3-8-1-6-geographic-position.html
+                    e = e
+                        .append_property(Property::new("LOCATION", &addr).done())
+                        .done();
+                }
+            }
+
+            if let Some(contact) = contact_from_ev(&ev) {
+                // https://icalendar.org/iCalendar-RFC-5545/3-8-4-2-contact.html
+                e = e
+                    .append_property(Property::new("CONTACT", &contact).done())
+                    .done();
+            }
+            if let Some(end) = ev.end {
+                e = e.ends(end).done();
+            }
+            if let Some(desc) = ev.description {
+                e = e.description(&desc).done();
+            }
+            e
+        })
+        .collect::<Calendar>();
+    let calendar = calendar.name("OpenFairDB events").done();
+    let data = calendar.to_string();
+    let c_type = ContentType::new("text", "calendar");
+    Ok(Content(c_type, data))
+}
+
+fn address_from_location(l: &Location) -> Option<String> {
+    l.address.as_ref().and_then(|a| {
+        if a.is_empty() {
+            None
+        } else {
+            let mut details = vec![];
```

Comment:
How about `once(a.street).chain(once(zip.clone())...filter(identity).collect()`? I always strive to avoud `mut` keyword if possible.

Whenever you convert a reference into an owned value you should prefer `to_owned()` instead of `clone()`, because this would be semantically correct.